pkg/result/result_helper.go

```diff
@@ -1,20 +1,54 @@
 package result
 
 import (
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 	"time"
 
 	ctrl "sigs.k8s.io/controller-runtime"
 )
 
-// Copyright DataStax, Inc.
-// Please see the included license file for details.
-
+// This is just so that we can reference TerminalError in the Godoc of [Error]
+var _ error = reconcile.TerminalError(nil)
+
+// ReconcileResult represents the result of a step in the reconciliation process.
+//
+// We typically split the top-level Reconcile() method of a controller into multiple sub-functions. Each of these
+// functions uses ReconcileResult to communicate to its caller how the current iteration should proceed. There are 4
+// possible implementations: [Continue], [Done], [RequeueSoon], and [Error].
+//
+// The idiomatic way to handle a ReconcileResult in an intermediary sub-function is:
+//
+//	if recResult := callStep1(); recResult.Completed() {
+//	   // Global success, requeue or error: stop what we're doing and propagate up
+//	   return recResult
+//	}
+//	// Otherwise, proceed with the next step(s)
+//	if recResult := callStep2(); recResult.Completed() {
+//	   // etc...
+//
+// The idiomatic way to handle a ReconcileResult in the top-level Reconcile() method is:
+//
+//	recResult := callSomeSubmethod()
+//	// Possibly inspect the result (e.g. to set an error field in the status)
+//	return recResult.Output()
 type ReconcileResult interface {
+	// Completed indicates that this result is *NOT* a [Continue].
```

Comment:
Issue: Defining something in terms of what it is not is a large part of our issue here I think. Can we define `Completed()` as what it is, rather than what it is not? Would it be clearer if we renamed it to `BailToMainReconciliation()` or something like that? Is that what `Completed()` really means?

Reply:
Exactly.
- loop: the entire process, from the moment we receive the first request to the moment we consider the object has reached the expected state.
- iteration: one call to the top-level `Reconcile()` method. One loop may consist of multiple iterations.
- step: an action in an iteration. Maps to sub-function calls for the sake of this API.

`Done` or `Error(terminalError)` ends the loop.
`Requeue` or `Error(nonTerminalError)` ends the current iteration.
`Continue` proceeds to the next step.

Reply:
> And this isn't a place to start discussing Done/Completed/Concluded/Finished/Ended

This is exactly the place to do so IMO because that's what we're having issues with. As I've said before, I would have serious concerns about any iteration of this PR which does not disambiguate between `Completed` and `Done` since they mean exactly the same thing in plain english and this is unclear from the way that these concepts are named in the codebase. 

> loop: the entire process, from the moment we receive the first request to the moment we consider the object has reached the expected state.
> iteration: one call to the top-level Reconcile() method. One loop may consist of multiple iterations.
> step: an action in an iteration. Maps to sub-function calls for the sake of this API.

While these concepts might form a useful basis for a future framework, and I think that bears discussion, I will once again re-iterate that they do not currently exist in the codebase. Also, once again, the design of a potential future framework is a matter for a design doc, not a PR.

Reply:
> And this isn't a place to start discussing Done/Completed/Concluded/Finished/Ended

It's ok to discuss the names and agree on their meaning, which doesn't mean we're changing those at the moment. Getting a common understanding is the first step to drive some changes that would clarify things going forward.

I agree with Miles that we shouldn't define something as what it is not, but rather as what it is.
@olim7t, could you maybe rework this comment by removing the first line, and also `In other words, it means that`. The rest of your comment looks good to me.

Reply:
I agree too, but I think it's more of a reflection about the design rather than the wording of the comment. "Not continue" is an accurate description when you consider the idiomatic usage of this method: `if recResult.Completed() { return recResult }` -- "if I've been asked not to continue, return now".

Anyway I'll switch the two sentences if that helps.